### PR TITLE
test: Add test for rpcwhitelistdefault

### PR DESCRIPTION
**Pull request by cvengler**
A test for `rpcwhitelistdefault`.

### CODE REVIEW

File: test/functional/rpc_whitelistdefault.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+)
+import http.client
+import urllib.parse
+
+def rpccall(node, user, password, method):
+    url = urllib.parse.urlparse(node.url)
+    headers = {"Authorization": "Basic " + str_to_b64str('{}:{}'.format(user, password))}
+    conn = http.client.HTTPConnection(url.hostname, url.port)
+    conn.connect()
+    conn.request('POST', '/', '{"method": "' + method + '"}', headers)
```
Comment (promag) [inline test/functional/rpc_whitelistdefault.py]: nit, use `.format(method)`.

File: test/functional/rpc_whitelistdefault.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""
+A test for nrpcwhitelistdefault states.
```
Comment (promag) [inline test/functional/rpc_whitelistdefault.py]: typo.

File: test/functional/rpc_whitelistdefault.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+import os
+from test_framework.util import (
+    get_datadir_path,
```
Comment (promag) [inline test/functional/rpc_whitelistdefault.py]: nit, sort

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by practicalswift** [comment]
Concept ACK

File: test/functional/rpc_whitelistdefault.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    conn = http.client.HTTPConnection(url.hostname, url.port)
+    conn.connect()
+    conn.request('POST', '/', '{{"method": "{}"}}'.format(method), headers)
+    resp = conn.getresponse()
+    conn.close()
+    return resp.code
+
+class RPCWhitelistDefaultTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 4
```
Comment (jonatack) [inline test/functional/rpc_whitelistdefault.py]: perhaps add: `self.setup_clean_chain = False` for clarity

**Review by jonatack — COMMENTED** [review]
I think this test should be added to the existing rpc_whitelist.py test with which this test shares the same imports and rpccall code.

**Comment by emilengler** [comment]
> I think this test should be added to the existing rpc_whitelist.py test with which this test shares the same imports and rpccall code.

They are two different options. The test would be to big to test both.

File: test/functional/rpc_whitelistdefault.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""
+A test for rpcwhitelistdefault states.
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+import os
```
[the remark this answers is not in the document]
Comment (emilengler) [reply test/functional/rpc_whitelistdefault.py]: Done

File: test/functional/rpc_whitelistdefault.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+from test_framework.util import (
+    assert_equal,
+    get_datadir_path,
+    str_to_b64str
```
[the remark this answers is not in the document]
Comment (emilengler) [reply test/functional/rpc_whitelistdefault.py]: The comma is at the end of every line or am I getting it wrong?

File: test/functional/rpc_whitelistdefault.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        self.log.info("Testing rpcwhitelistdefault=1 with no specififed permissions")
+        assert_equal(403, rpccall(self.nodes[2], "user1", "12345", "getblockcount"))
+        assert_equal(403, rpccall(self.nodes[2], "user2", "12345", "getblockcount"))
+        self.log.info("Testing rpcwhitelistdefault=1 with specififed permissions")
+        assert_equal(200, rpccall(self.nodes[3], "user1", "12345", "getbestblockhash"))
+        assert_equal(403, rpccall(self.nodes[3], "user2", "12345", "getbestblockhash"))
+
+if __name__ == "__main__":
+    RPCWhitelistDefaultTest().main()
+
```
[the remark this answers is not in the document]
Comment (emilengler) [reply test/functional/rpc_whitelistdefault.py]: The linter cries if there is no new line ant the end of the file. See PEP 8

File: test/functional/rpc_whitelistdefault.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            f.write("rpcwhitelistdefault=1\n")
+            f.write("rpcauth=user1:50358aa884c841648e0700b073c32b2e$b73e95fff0748cc0b517859d2ca47d9bac1aa78231f3e48fa9222b612bd2083e\n")
+            f.write("rpcauth=user2:6e0499f40b6420da2d6a3baaec1a1268$4c03b56943444ad141e3a2d8389aa22b2ade5b09d6710d5299ade7b2902086f1\n")
+            f.write("rpcwhitelist=user1:getbestblockhash\n")
+            # We need to permit the __cookie__ user for the Bitcoin Test Framework
+            f.write("rpcwhitelist=__cookie__:getblockcount,getwalletinfo,importprivkey,getblockchaininfo,submitblock,addnode,getpeerinfo,getbestblockhash,getrawmempool,syncwithvalidationinterfacequeue,stop\n")
+
+    def run_test(self):
+        self.log.info("Testing rpcwhitelistdefault=0 with no specififed permissions")
+        assert_equal(200, rpccall(self.nodes[0], "user1", "12345", "getbestblockhash"))
```
[the remark this answers is not in the document]
Comment (emilengler) [reply test/functional/rpc_whitelistdefault.py]: This is a featrue for another PR. Other tests use the plain codes as well

File: test/functional/rpc_whitelistdefault.py (hunk 4 above)
Comment (emilengler) [reply test/functional/rpc_whitelistdefault.py]: It doesn't matter anyway

Comment (jonatack) [reply test/functional/rpc_whitelistdefault.py]: It is false by default (see test_framework.py::L96), but most tests set it explicitly even when false. It matters a bit for test speed as it controls whether or not to use the cached data directories for the test setup.

File: test/functional/rpc_whitelistdefault.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/rpc_whitelistdefault.py]: Have a look at example_test.py

**Comment by jonatack** [comment]
> > I think this test should be added to the existing rpc_whitelist.py test with which this test shares the same imports and rpccall code.
> 
> They are two different options. The test would be to big to test both.

The tests can be organised into different methods within the same test class and file. Many functional test files are much larger than this one would become. Have a look at the size at p2p_segwit.py, feature_block.py, etc. It may be a bit faster in test run time to group similar tests into the same testfile setup where possible. And otherwise, rpccall could be extracted to a common helper method (if one does not already exist). I could be wrong but this is what I'd try to do. It's up to you.

**Comment by emilengler** [comment]
Even if the featrue is very similar, they test both two different things. `rpc_whitelist.py` tests permissions, invalid syntaxes, etc.
This test tests if users can even do something but is not limited to specific permissions

**Review by promag — COMMENTED** [review]
I think the new file is fine.

However have you considered to use just one node? Like
 1. change config
 2. restart
 3. test calls/access
 4. repeat

I prefer this way because:
 - new test cases don't imply messing with more nodes;
 - testing the calls/access is immediately after the config.

**Comment by emilengler** [comment]
> However have you considered to use just one node? Like

Yes but after all I believe this would make the code much messier because I would need to re-implement some functions. This is probably the most cleanest way on how to do it

**Comment by fanquake** [comment]
Closing as "Up for grabs". If these tests are going to be added, I think adding them to the existing `rpc_whitelist.py` is the better approach, as well as using a single node as suggested.

**Pull request closed without merging**